1.MLP_Binary_Classifier/MLP_utils/utils.py

```diff
@@ -0,0 +1,627 @@
+            it is nice to know about what the other output metrics are
+
+    Return:
+        metric(s) : str or float
+            if return_info == True:
+                Mean Validation Accuracy
+                Mean Validation Loss
+                Mean Training Accuracy
+                Mean Training Loss
+            if return_info == False:
+                return the mean validation accuracy
+
+    """
+
+    # calling model function
+    model = build_model_custom(trial, in_features)
+
+    # param dictionary for optimization
+    optimization_params = {
+        "learning_rate": trial.suggest_float("learning_rate", 1e-5, 1e-1),
+        "optimizer": trial.suggest_categorical("optimizer", ["Adam", "RMSprop", "SGD"]),
+        "n_unit": trial.suggest_int("n_unit", 1, 50),
+    }
+
+    # param optimizer pick
+    optimizer = getattr(optim, optimization_params["optimizer"])(
+        model.parameters(), lr=optimization_params["learning_rate"]
+    )
+    # loss function
+
+    # for binary model use different for multi-class
+    criterion = nn.BCEWithLogitsLoss()
+
+    # send model to device(cuda)
+
+    model = model.to(params.DEVICE)
+    criterion = criterion.to(params.DEVICE)
+
+    # train set accuracy and loss
+    train_acc = []
+    train_loss = []
+
+    # validation set accuracy and loss
+    valid_acc = []
+    valid_loss = []
+
+    # total number of data to pass through
+    total_step = len(train_loader)
+    total_step_val = len(valid_loader)
+
+    for epoch in range(params.OPTIM_EPOCHS):
+
+        (
+            train_acc,
+            train_loss,
+            valid_acc,
+            valid_loss,
+            correct,
+            total,
+            correct_v,
+            total_v,
+        ) = train_n_validate(
+            model,
+            optimizer,
+            criterion,
+            train_acc,
+            train_loss,
+            valid_acc,
+            valid_loss,
+            total_step,
+            total_step_val,
+            params,
+            train_loader,
+            valid_loader,
+        )
+
+        trial.report(np.mean(valid_loss), epoch)
+
+        # Handle pruning based on the intermediate value
+        if trial.should_prune():
+            raise optuna.exceptions.TrialPruned()
+
+    # I want information returned but only 1 metric required for the optimize function called by study.optimize
+    # with out this conditional statement the optimization will fail
+    if return_info == True:
+        print(f"Validation Accuracy: {np.mean(valid_acc)}")
+        print(f"Validation Loss: {np.mean(valid_loss)}")
+        print(f"Training Accuracy: {np.mean(train_acc)}")
+        print(f"Training Loss: {np.mean(train_loss)}")
+    else:
+        return np.mean(valid_loss)
+
+
+def extract_best_trial_params(best_params):
+    """
+    This function extractions the best parameters from the best trial.
+
+    These extracted parameters will be used to create a new model.
```

Comment:
```suggestion
    We will use these extracted parameters to train a new model.
```

Please use active voice instead of passive voice! :)